Editor/Screenshots: New - capture screenshot on publish

Nevoss opened the pull request:
## PR Checklist
<!-- 
Please check if your PR fulfills the following requirements:
**Filling out the template is required.** Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
 -->
- [x] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md


## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x" with no spaces eg: [x]. -->
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Description
An explanation of what is done in this PR

* On publish it creates a screenshot of the current view of the post

## Test instructions
This PR can be tested by following these steps:

* Open a post with the editor
* Make some changes as you wish
* Publish the post
* Make sure that screenshot is created at the upload directory
* Make sure that the screenshot look the same as the preview of the post.

## Quality assurance

- [x] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

matipojo committed:
new module: screenshots

matipojo committed:
screenshots: use  rasterizeHTML library & an hidden iframe to take the screenshot

matipojo committed:
Merge branch 'release/2.1.7' into feature/screenshots

# Conflicts:
#	assets/js/editor.js
#	assets/js/editor.min.js
#	core/modules-manager.php
#	includes/controls/groups/image-size.php

matipojo committed:
Screenshots: Progress.

matipojo committed:
Screenshots: New - Add YouTube iframe fallback

matipojo committed:
Merge branch 'feature/screenshots' of https://github.com/matipojo/elementor into feature/screenshots

matipojo committed:
Merge branch 'feature/screenshots' of https://github.com/matipojo/elementor into feature/screenshots

matipojo committed:
Merge branch 'release/2.2.8'

matipojo committed:
Merge branch 'release/2.4.0'

matipojo committed:
Screenshots: Fix - Use `html2canvas` library, Common.Ajax + other fixes.

matipojo committed:
Merge branch 'release/2.4.0' of https://github.com/pojome/elementor into feature/screenshots

matipojo committed:
Merge remote-tracking branch 'elementor/release/3.0.0' into feature/screenshots

matipojo committed:
Merge remote-tracking branch 'elementor/release/3.0.0' into feature/screenshots

Nevoss committed:
merge fixes

Nevoss committed:
Editor/Screenshots: Change the implamentation to Js API

Nevoss committed:
Editor/Screenshots: Change html2canvas library to updated version.

Nevoss committed:
Editor/Screenshot: Change screenshot main js file to modern syntax.

Nevoss committed:
Editor/Screenshots: Fix images and fonts renders in the screenshot + image margin.

Nevoss committed:
Editor/Screenshots: Update html2canvas library and use SCRIPT_DEBUG const

Nevoss committed:
Editor/Screenshot: Change html2canvas with dom-to-image

Nevoss committed:
Editor/Screenshots: Fix font loading

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/elementor/elementor?pullRequest=11751) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/elementor/elementor?pullRequest=11751) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: matipojo<br/>:x: Nevoss<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/elementor/elementor?pullRequest=11751) it.</sub>

Nevoss committed:
Editor/Screenshots: Fix - fonts loaded

matipojo commented on modules/screenshots/assets/js/frontend/screenshot.js:
```diff
@@ -0,0 +1,267 @@
+		}
+
+		return Promise.resolve()
+			.then( this.handleIFrames.bind( this ) )
```
Try remove the promise if not needed.

matipojo commented on modules/screenshots/assets/js/frontend/screenshot.js:
```diff
@@ -0,0 +1,267 @@
+	}
+
+	/**
+	 * The html-to-image library can not snapshot the iframes.
```
```suggestion
	 * Html-to-image libraries can not snapshot the iframes.
```

matipojo commented on modules/screenshots/assets/js/frontend/screenshot.js:
```diff
@@ -0,0 +1,267 @@
+	handleIFrames() {
+		const $self = this;
+
+		this.$elementor.find( 'iframe' ).each( function () {
```
Try use arrow function + remove the $self.

matipojo commented on modules/screenshots/assets/js/frontend/screenshot.js:
```diff
@@ -0,0 +1,267 @@
+	}
+
+	/**
+	 * css from another server cannot be loaded.
```
```suggestion
	 * CSS from another server cannot be loaded.
```

matipojo commented on modules/screenshots/assets/js/frontend/screenshot.js:
```diff
@@ -0,0 +1,267 @@
+
+	/**
+	 * css from another server cannot be loaded.
+	 * that is why we fetch the content of the css and set it as an inline css.
```
```suggestion
	 * that is why it needed to be fetch the content of the css and set it as an inline css.
```

matipojo commented on modules/screenshots/assets/js/frontend/screenshot.js:
```diff
@@ -0,0 +1,267 @@
+	 * @returns {Promise<unknown[]>}
+	 */
+	loadExternalCss() {
+		const $self = this;
```
Try remove.

matipojo commented on modules/screenshots/assets/js/frontend/screenshot.js:
```diff
@@ -0,0 +1,267 @@
+	createImage() {
+		return new Promise( ( resolve ) => {
+			setTimeout( () => {
+				domtoimage.toPng( this.$elementor.parents( 'body' ).get( 0 ), {} )
```
`document.body`.

matipojo commented on modules/screenshots/module.php:
```diff
@@ -0,0 +1,150 @@
+	}
+
+	public function enqueue_scripts() {
+		if ( ! isset( $_REQUEST['elementor-screenshot'] ) || ! User::is_current_user_can_edit() ) { // WPCS: CSRF ok.
```
```suggestion
		if ( ! isset( $_REQUEST['elementor-screenshot'] ) || ! User::is_current_user_can_edit() ) {
```

Nevoss committed:
Editor/Screenshots: Fix - Aesthetic changes based on the code review

Nevoss committed:
Editor/Screenshots: Fix - move assets to right dir + format code little bit

Nevoss committed:
Editor/Screenshots: Change - screenshot url is now attached to the document config.

Nevoss committed:
Editor/Screenshots: Change - format + add some decription to capture screenshot hook

Nevoss committed:
Editor/Screenshot: Change - implamentation of screenshot on the front side

Nevoss committed:
Editor/Screenshots: Fix - load img in screenshot context

Nevoss committed:
Editor/Screenshots: Chnage - Aesthetic changes.

Nevoss committed:
Editor/Screenshots: Change - revert render mode feature on Fronend.

Nevoss committed:
Editor/Screenshots: Change - creates a screenshot when the document fully loaded or after 5 secs

Nevoss committed:
Editor/Screenshots: Chnage - better logs

Nevoss committed:
Editor/Screenshot: New - alert when tring to close window before screenshot was taken.

Nevoss committed:
Editor/Screenshots: Chnage - typo fix + comments

matipojo commented on modules/screenshots/assets/js/editor/component.js:
```diff
@@ -0,0 +1,40 @@
+		return this.importHooks( hooks );
+	}
+
+	isCapturingScreenshot() {
```
It's a public property, so it don't needs a getter.

matipojo commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -0,0 +1,315 @@
+	 * @returns {Promise<unknown[]>}
+	 */
+	loadExternalCss() {
+		const selector = this.config.allowedCssUrls.map( ( allowedCssUrl ) => {
```
What about adding a proxy URL to the original external link?

matipojo commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -0,0 +1,315 @@
+		return fetch( url )
+			.then( ( response ) => response.text() )
+			.then( ( data ) => {
+				// This is a specific use case when we need
```
We :)

matipojo commented on modules/screenshots/module.php:
```diff
@@ -0,0 +1,180 @@
+	}
+
+	public function screenshot_proxy() {
+		if ( ! wp_verify_nonce( $_GET['nonce'], 'screenshot_proxy' ) ) {
```
Do we have a pattern for allowed URLs? (like `/(png|jpg|css)/`)

matipojo commented on modules/screenshots/module.php:
```diff
@@ -0,0 +1,180 @@
+		$post_id = $data['post_id'];
+
+		$file_content = substr( $data['screenshot'], strlen( 'data:image/png;base64,' ) );
+		$file_name = 'Elementor Post Screenshot ' . $post_id . '.png';
```
I think your preview fix was good:
```suggestion
		$file_name = 'Elementor-post-screenshot-' . $post_id . '.png';
```

matipojo commented on modules/screenshots/module.php:
```diff
@@ -0,0 +1,180 @@
+
+		add_filter( 'wp_unique_filename', $over_write_file_name_callback );
+
+		$upload = wp_upload_bits(
```
Try move all screenshots to a new folder `/wp-content/uploads/elementor/screenshots`. (create if not exist with `wp_mkdir_p`)

matipojo commented on modules/screenshots/module.php:
```diff
@@ -0,0 +1,180 @@
+	}
+
+	public function enqueue_scripts() {
+		if ( ! $this->is_in_screenshot_mode() || ! User::is_current_user_can_edit() ) {
```
```suggestion
		if ( ! $this->is_screenshot_mode() || ! User::is_current_user_can_edit() ) {
```

Nevoss committed:
Editor/Screenshots: Change - var names and comments.

Nevoss committed:
Editor/Screenshots: Change - proxy external css files

Nevoss committed:
Editor/Screenshots: New - timeout for screenshot

Nevoss committed:
Editor/Screenshots: New - timeout for screenshot

Nevoss committed:
Editor/Screenshots: Change - screenshot upload folder.

matipojo commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -0,0 +1,329 @@
+		if ( ! this.$elementor.length ) {
+			elementorCommon.helpers.consoleWarn( 'Screenshots: Elementor content was not found.' );
+
+			return;
```
Run `screenshotFailed` in order to clear the timer.

matipojo commented on modules/screenshots/assets/js/editor/commands/internal/capture.js:
```diff
@@ -0,0 +1,35 @@
+				return;
+			}
+
+			this.component.isCapturingScreenshot = false;
```
Remove the iframe element.

Nevoss committed:
Editor/Screenshots: New - remove the iframe after screenshot

Nevoss committed:
Merge branch 'release/3.0.0' into feature/screenshots

Nevoss committed:
Editor/Screenshots: Change - remove unnecessary element before taking screenshot.

Nevoss committed:
Editor/Screenshots: Fix - height of the the image is max-height and not fixed height.

Nevoss committed:
Editor/Screenshots: Fix - typo

Nevoss committed:
Editor/Screenshot: New - comments in module class.

Nevoss committed:
Editor/Screenshots: Change - disable the creation of thumbnail sizes.

Nevoss committed:
Editor/Screenshots: New - Test [WIP]

Nevoss committed:
Editor/Screenshots: New - more tests.

Nevoss committed:
Editor/Screenshots: Fix - remove new phpunit test methods.

Nevoss committed:
Editor/Screenshots: Change - refactor to more testable module.

matipojo committed:
Module/Screenshots: Fix - Cannot set mode for 3rd party document types.

Set mode on template redirect after plugins are loaded and before WP page render.

matipojo committed:
Module/Screenshots: Fix - Cannot set mode for 3rd party document types.

Set mode on template redirect after plugins are loaded and before WP page render.

Nevoss committed:
Editor:Screenshots: Change - Refactor screenshot config method to 2 methods for outer usage.

Nevoss committed:
Module\Screenshots: Chnage - make it work with the site editor.

matipojo committed:
Merge remote-tracking branch 'Nevoss/feature/screenshots' into feature/screenshots

matipojo committed:
Modules/Screenshots: Fix - Assets file not loaded after build.

All JS files outside the `assets` folder - are deleted during the build process.

Nevoss committed:
Merge pull request #2 from matipojo/feature/screenshots

Feature/screenshots - Fixes.

Nevoss committed:
Merge branch 'feature/render-mode' into feature/screenshots

Nevoss committed:
Module/Screenshots: New - Interate screenshot with render mode.

Nevoss committed:
Module/Screenshots: Change - File naming + index.html at screenshot dir + bundle all the upload to sperate class.

Nevoss committed:
Module/Screenshots: New - generate empty screenshot for eampty posts.

Nevoss committed:
Module/Screenshots: New - test the command itself.

Nevoss committed:
Module/Screenshots: Fix - remove the "only" from the tests.

Nevoss committed:
Module/Screenshots: Fix - test.

Nevoss committed:
Module/Screenshots: Change - resolve command if succeed and reject if failed.

matipojo commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -0,0 +1,378 @@
+				width: 1200,
+				height: 1500,
+			},
+			excludedCssUrls: [
+				'https://kit-pro.fontawesome.com',
+			],
+			timeout: 15000, // Wait until screenshot taken or fail in 15 secs.
+			render_timeout: 5000, // Wait until all the element will be loaded or 5 sec and then take screenshot.
+			timerLabel: null,
```
Keep all config keys with `snake_case` format.

matipojo commented on modules/screenshots/screenshot.php:
```diff
@@ -0,0 +1,199 @@
+<?php
+
```
Remove all top empty lines.

matipojo commented on modules/screenshots/screenshot.php:
```diff
@@ -0,0 +1,199 @@
+	 */
+	public function get_file_name() {
+		if ( ! $this->file_name ) {
+			$now = ( new \DateTime() )->format( 'Y-m-dTH:i:s' );
```
Use valid file name characters.
```suggestion
			$now = ( new \DateTime() )->format( 'Y-m-d-H-i-s' );
```

matipojo commented on tests/phpunit/elementor/modules/screenshots/module/test-ajax-action.php:
```diff
@@ -0,0 +1,62 @@
+	}
+
+	public function test_should_return_false_if_there_is_no_post_id() {
+		$response = $this->module->ajax_save( [ 'screenshot' => 'adsasd' ] );
```
```suggestion
		$response = $this->module->ajax_save( [ 'screenshot' => 'invalid-id' ] );
```

Nevoss committed:
Module/Screenshots: Change - Code review changes.

Nevoss committed:
Merge branch 'release/3.0.0' into feature/screenshots

# Conflicts:
#	assets/dev/js/editor/editor-base.js
#	tests/qunit/tests.js

Nevoss committed:
Modules:Screenshot - Fix - remove types for php 5.6

Nevoss committed:
Module:Screenshots: Fix - test image should attached to post.

Nevoss committed:
Module:Screenshot - Fix - code style.

matipojo commented on tests/phpunit/elementor/modules/screenshots/module/test-ajax-action.php:
```diff
@@ -0,0 +1,64 @@
+<?php
+
```
Remove all top empty lines.

Nevoss committed:
Merge branch 'release/3.0.0' into feature/screenshots

# Conflicts:
#	.grunt-config/webpack.js
#	tests/qunit/tests.js

Nevoss committed:
Module:Screenshot - Change - code style.

matipojo commented on modules/screenshots/assets/js/editor/commands/internal/capture.js:
```diff
@@ -0,0 +1,57 @@
+		// then the command send a notice to the component about it.
+		return new Promise( ( resolve, reject ) => {
+			const listener = ( message ) => {
+				const { data } = message;
+
+				if ( ! data.name || data.name !== 'capture-screenshot-done' ) {
+					return;
+				}
+
+				this.component.isCapturingScreenshot = false;
+
+				if ( ! elementorCommonConfig.isDebug ) {
+					this.iframe.remove();
+				}
+
+				window.removeEventListener( 'message', listener );
+
+				return data.success ?
+					resolve( this.iframe ) :
+					reject();
+			};
+
+			window.addEventListener( 'message', listener );
+		} );
+	}
+
+	createIframe() {
+		const iframe = document.createElement( 'iframe' );
+
+		iframe.src = this.getIframeUrl();
+		iframe.width = '1200';
+		iframe.style = 'visibility: hidden;';
+
+		return iframe;
+	}
+
+	getIframeUrl() {
+		return elementor.config.document.urls.screenshot;
+	}
+}
```
It's not needed here.

kobizz replied:
Our standard is to use jQuery for dom manipulations and queries

matipojo commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -0,0 +1,378 @@
+/* global ElementorScreenshotConfig, jQuery */
+class Screenshot {
```
It's not a part of the editor environment. it's more like a "frontend handle" bu the handlers system is not available on static render.

kobizz replied:
Our standard is to handle all our views modules the same way. There is a ViewModule base available on both ends and you should use it.

Nevoss committed:
Merge branch 'release/3.0.0' of github.com:elementor/elementor into feature/screenshots

Nevoss committed:
Module : Screenshot - Fix - Remove filters that already handled in the server and handle youtube thumbnail.

Nevoss commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -0,0 +1,378 @@
+	 */
+	handleYouTubeIFrame( $iframe, $iframeMask ) {
+		const regex = /^.*(?:youtu.be\/|youtube(?:-nocookie)?.com\/(?:(?:watch)??(?:.*&)?vi?=|(?:embed|v|vi|user)\/))([^?&"'>]+)/;
+		const matches = regex.exec( $iframe.attr( 'src' ) );
```
Function removed from the code (not needed any more).

Nevoss commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -0,0 +1,378 @@
+		this.$elementor.find( '.elementor-slides' ).each( ( index, el ) => {
+			const $this = jQuery( el );
+
+			$this.find( '> *' ).not( $this.find( '> :first-child' ) ).each( ( childIndex, childEl ) => {
```
Function removed from the code (not needed any more).

Nevoss commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -0,0 +1,378 @@
+		const regex = /url\(\"(.*)\"/gm;
+		const url = $iframe.next().css( 'backgroundImage' );
+		const matches = regex.exec( url );
+
```
Function removed from the code (not needed any more).

Nevoss committed:
Module : Screenshot - Fix - Code style.

matipojo commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -87,86 +92,6 @@ class Screenshot {
-	 * Html to images libraries can not snapshot IFrames
-	 * this method convert all the IFrames to some other elements.
-	 */
-	handleIFrames() {
```
We still need it for custom iframes.

matipojo commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -192,6 +117,21 @@ class Screenshot {
+	 * Make a proxy to images urls that has some problems with cross origin (like youtube).
+	 */
+	loadExternalImages() {
+		const selector = this.config.include_images_urls
```
```suggestion
		const selector = this.config.external_images_urls
```

Nevoss committed:
Module : Screenshot - Fix - Handle iFrames.

Nevoss committed:
Merge branch 'release/3.0.0' into feature/screenshots

# Conflicts:
#	assets/dev/js/editor/editor-base.js

Nevoss committed:
Module : Screenshot - Fix - Safari browser support

Safari will execute screenshot with html2canvas, all the other browsers will use dom-to-image.

Nevoss committed:
Module : Screenshot - Change - Check fo safari

Nevoss committed:
Merge branch 'release/3.0.0' of github.com:elementor/elementor into feature/screenshots

Nevoss committed:
Module : Screenshots - Change - use jquery in capture command

Nevoss committed:
Module : Screenshots - Change - extend base ViewModule

kobizz commented on assets/lib/html2canvas/js/html2canvas.min.js:
```diff
@@ -0,0 +1,20 @@
+/*!
```
This library weights 120k in its minified version. I don't think it worths it.

kobizz commented on modules/screenshots/assets/js/editor/commands/internal/capture.js:
```diff
@@ -0,0 +1,59 @@
+
+		this.$iframe = this.createIframe();
+
+		jQuery( 'body' ).append( this.$iframe );
```
```suggestion
		elementorCommon.elements.$body.append( this.$iframe );
```

kobizz commented on modules/screenshots/assets/js/editor/commands/internal/capture.js:
```diff
@@ -0,0 +1,59 @@
+	}
+
+	createIframe() {
+		const $iframe = jQuery( '<iframe></iframe>' );
+
+		$iframe.attr( 'src', this.getIframeUrl() );
```
You can simplify it:

```suggestion
		const $iframe = jQuery( '<iframe>', { src: this.getIframeUrl() } );
```

kobizz commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -0,0 +1,352 @@
+	 * Hide all the element except for the target element.
+	 */
+	hideUnnecessaryElements() {
+		jQuery( 'body' ).prepend(
```
```suggestion
		elementorCommon.elements.$body.prepend(
```

kobizz commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -0,0 +1,352 @@
+			this.elements.$elementor
+		);
+
+		jQuery( 'body > *' ).not( this.elements.$elementor ).css( 'display', 'none' );
```
Consider caching `jQuery( 'body > *' ).not( this.elements.$elementor )` inside `getDefaultElements`

kobizz commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -0,0 +1,352 @@
+
+			$newLink.attr( 'href', this.getScreenshotProxyUrl( $link.attr( 'href' ) ) );
+
+			jQuery( 'head' ).append( $newLink );
```
cache `jQuery( 'head' )` inside `getDefaultElements`

kobizz commented on modules/screenshots/assets/js/preview/screenshot.js:
```diff
@@ -0,0 +1,352 @@
+	removeUnnecessaryElements() {
+		let currentHeight = 0;
+
+		this.elements.$elementor
+			.find( ' .elementor-section-wrap > .elementor-section' )
```
Consider caching it inside `getDefaultElements`

kobizz commented on modules/screenshots/screenshot.php:
```diff
@@ -0,0 +1,198 @@
+	 *
+	 * @return $this
+	 */
+	public function upload() {
```
Add a TODO to use the planned Uploads Manager here

Nevoss committed:
Modules : Screenshot - Change - Caching jQuery elements.

Nevoss committed:
Modules : Screenshot - Fix - Use elementor common body.

matipojo commented on modules/screenshots/module.php:
```diff
@@ -0,0 +1,250 @@
+
+		wp_add_inline_script( 'elementor-screenshot', 'var ElementorScreenshotConfig = ' . wp_json_encode( $config ) . ';' );
+
+		$css = Post_Preview::create( $post_id );
+		$css->enqueue();
```
Try remove.

Nevoss committed:
Modules : Screenshots - Change - Remove PostPreview

Nevoss committed:
Module : Screenshots - Fix - Handling errors in screenshot.

mark the screenshot as failed + set default dataURL when failed to load an image.

Nevoss committed:
Modules : Screenshot - Fix - Wait for mark as failed before remove the iFrame.
